bin/export_grafana_snapshot.py

```diff
@@ -0,0 +1,182 @@
+        url = self.construct_url(path)
+
+        print("Search dashboard with name {0}".format(dashboard_name))
+
+        resp = self.session.get(url)
+
+        if not resp.ok:
+            raise Exception(resp.content)
+        if len(resp.json()) == 0:
+            raise Exception("cannot find dashboard with name")
+        return resp.json()[0].get("uid")
+
+    def post_snapshot(self, dashboard, name="Istio Dashboard",
+                      expires=None, key=None, deletekey=None, external=False):
+        """
+        Export dashboard to snapshot
+        Snapshot name is formatted with dashboard name and timestamp
+        """
+        path = "api/snapshots"
+        if external is True:
+            if deletekey is None or key is None:
+                print("Please specify key and delete key for external snapshot")
+                exit(1)
+
+        data = dashboard
+        data['external'] = external
+        current_dt = datetime.datetime.now().strftime('%Y-%m-%d %H:%M')
+        data['name'] = name + "_" + str(current_dt)
+        if expires is not None:
+            data['expires'] = expires
+        if key is not None:
+            data['key'] = key
+        if deletekey is not None:
+            data['deletekey'] = deletekey
+
+        print("Export dashboard name: {0} as snapshot, key:{1}"
+              .format(data['name'], key))
+
+        resp = self.session.post(self.construct_url(path), json=data)
+
+        if not resp.ok:
+            raise Exception(resp.content)
+        return resp.json()
+
+    def get_snapshots(self):
+        """
+        :return: all the snapshots exported
+        """
+        path = "api/dashboard/snapshots"
+        url = self.construct_url(path)
+        print(url)
+
+        resp = self.session.get(url)
+        if not resp.ok:
+            raise Exception(resp.content)
+        print(json.dumps(resp.json(), indent=4, sort_keys=True))
+
+    def construct_url(self, path, host_url=None, port=None, protocol="http"):
+        """
+        To construct url for request
+        """
+        if host_url is None:
+            host_url = self.grafana_host['host']
+        if port is None:
+            port = self.grafana_host['port']
+        params = {
+            'protocol': protocol,
+            'host': host_url,
+            'path': path,
+        }
+        if port is None:
+            url_pattern = '{protocol}://{host}/{path}'
+        else:
+            params['port'] = port
+            url_pattern = '{protocol}://{host}:{port}/{path}'
+
+        return url_pattern.format(**params)
+
+    def export_snapshot(self, dashboard_name, external):
+        """
+        :param dashboard_name: dashboard to export snapshot
+        :param external: export locally or to raintank
+        """
+        dashboard_id = self.search_dashboard(dashboard_name)
+        dashboard =\
+            self.get_dashboard_by_uid(dashboard_id)
+        self.post_snapshot(dashboard,
```

Comment:
Return value is ingnored.
We need to know the uuid where this snapshot is exported so we can associated tags.

What is the persistence story for local exports?

Reply:
Ok, I guess you mean the url of where the snapshot exported? 

For local exports, it is persisted in local grafana db.

Do we still need another storage to store the exported snapshot information? Actually, there is api to query information of snapshots exported, also snapshot would be exported with name consist of dashboard name and timestamp or some other info.

Reply:
We can maintain a grafana instance with a persistent store, but the grafana we will be exporting from may be destroyed. So we cannot just keep the data there.
We should associate is the istio build number with this data.

Reply:
Please check the comment on top for update